### PR TITLE
`sage.calculus`: Modularization fixes, doctest cosmetics, `# needs`

### PR DESCRIPTION
**Pull request by mkoeppe**
### :books: Description
Not all of `sage.calculus` is for symbolic calculus. We change some import statements to `lazy_import` so that some modules can be loaded even when the symbolic subsystem is not present, or move them into methods.

We also add `# needs` tags to doctests to enable doctesting of modularized distributions.

- Part of: #29705 
- Cherry-picked from: #35095 

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/calculus/interpolation.pyx [hunk 1]
```diff
@@ -55,39 +55,39 @@ cdef class Spline:
 
     This example is in the GSL documentation::
 
-        sage: v = [(i + sin(i)/2, i+cos(i^2)) for i in range(10)]
-        sage: s = spline(v)
-        sage: show(point(v) + plot(s,0,9, hue=.8))
+        sage: v = [(i + sin(i)/2, i+cos(i^2)) for i in range(10)]                       # optional - sage.symbolic
```
Comment (orlitzky) [inline src/sage/calculus/interpolation.pyx]: These are all floats (passed through to GSL), so you can probably avoid this batch of `# optional` tags with `math.sin` and `math.cos`.

File: src/sage/calculus/interpolators.pyx [hunk 2]
```diff
@@ -117,7 +119,7 @@ cdef class PSpline:
             sage: ps = polygon_spline(pts)
             sage: ps.value(.5)
             (-0.363380227632...-1j)
-            sage: ps.value(0) - ps.value(2*pi)
+            sage: ps.value(0) - ps.value(2*pi)                                          # optional - sage.symbolic
```
Comment (orlitzky) [inline src/sage/calculus/interpolators.pyx]: This also takes a float as an argument, so `math.pi` avoids sage.symbolic.

File: src/sage/calculus/interpolation.pyx (hunk 1 above)
Comment (mkoeppe) [reply src/sage/calculus/interpolation.pyx]: Yes, there's a lot of code that does this in Sage; also in sage.plot.

I'll rewrite it. Thanks for taking a look at this in detail!

File: src/sage/calculus/interpolators.pyx [hunk 3]
```diff
@@ -179,15 +181,16 @@ def complex_cubic_spline(pts):
...
+        sage: m = Riemann_Map([lambda x: cs.value(real(x))],
+        ....:                 [lambda x: cs.derivative(real(x))], 0)
+        sage: show(m.plot_colored() + m.plot_spiderweb())                               # optional - sage.plot
 
     Polygon approximation of a circle::
 
-        sage: pts = [e^(I*t / 25) for t in range(25)]
-        sage: cs = complex_cubic_spline(pts)
-        sage: cs.derivative(2)
+        sage: pts = [e^(I*t / 25) for t in range(25)]                                   # optional - sage.symbolic
```
Comment (orlitzky) [inline src/sage/calculus/interpolators.pyx]: Same thing with `1j` for `I` and `math.e` for `e`.

File: src/sage/calculus/interpolators.pyx [hunk 4]
```diff
@@ -269,7 +272,7 @@ cdef class CCSpline:
             sage: cs = complex_cubic_spline(pts)
             sage: cs.value(4 / 7)
             (-0.303961332787...-1.34716728183...j)
-            sage: cs.value(0) - cs.value(2*pi)
+            sage: cs.value(0) - cs.value(2*pi)                                          # optional - sage.symbolic
```
Comment (orlitzky) [inline src/sage/calculus/interpolators.pyx]: Down here too

File: src/sage/calculus/ode.pyx [hunk 5]
```diff
@@ -289,7 +289,7 @@ class ode_solver():
         sage: T = ode_solver(g_1,y_0=[0,1,1],scale_abs=[1e-4,1e-4,1e-5],error_rel=1e-4, algorithm="rk8pd")
         sage: T.ode_solve(t_span=[0,12],num_points=100)
         sage: f = T.interpolate_solution()
-        sage: f(pi)
+        sage: f(pi)                                                                     # optional - sage.symbolic
```
Comment (orlitzky) [inline src/sage/calculus/ode.pyx]: And here

File: src/sage/calculus/transforms/dft.py [hunk 6]
```diff
@@ -412,11 +423,11 @@ def dst(self):
         EXAMPLES::
 
             sage: J = list(range(5))
-            sage: I = CC.0; pi = CC(pi)
-            sage: A = [exp(-2*pi*i*I/5) for i in J]
-            sage: s = IndexedSequence(A,J)
+            sage: I = CC.0; pi = CC(pi)                                                 # optional - sage.symbolic
```
Comment (orlitzky) [inline src/sage/calculus/transforms/dft.py]: This works with `math.pi` too

File: src/sage/calculus/transforms/dft.py [hunk 7]
```diff
@@ -390,9 +401,9 @@ def dct(self):
         EXAMPLES::
 
             sage: J = list(range(5))
-            sage: A = [exp(-2*pi*i*I/5) for i in J]
-            sage: s = IndexedSequence(A,J)
-            sage: s.dct()
+            sage: A = [exp(-2*pi*i*I/5) for i in J]                                     # optional - sage.symbolic
```
Comment (orlitzky) [inline src/sage/calculus/transforms/dft.py]: This one really uses the symbolic `pi` and `e`, but could easily be rewritten to use some other familiar sequence like the triangular numbers

File: src/sage/calculus/transforms/dwt.pyx [hunk 8]
```diff
@@ -63,27 +63,27 @@ def WaveletTransform(n, wavelet_type, wavelet_k):
...
     This example gives a simple example of wavelet compression::
 
         sage: a = DWT(2048,'daubechies',6)
-        sage: for i in range(2048): a[i]=float(sin((i*5/2048)**2))
-        sage: a.plot().show()  # long time (7s on sage.math, 2011)
-        sage: a.forward_transform()
-        sage: for i in range(1800): a[2048-i-1] = 0
-        sage: a.backward_transform()
-        sage: a.plot().show()  # long time (7s on sage.math, 2011)
+        sage: for i in range(2048): a[i]=float(sin((i*5/2048)**2))                      # optional - sage.symbolic
```
Comment (orlitzky) [inline src/sage/calculus/transforms/dwt.pyx]: `float(sin(x))` can be `math.sin(x)`

File: src/sage/calculus/transforms/dft.py (hunk 6 above)
Comment (mkoeppe) [reply src/sage/calculus/transforms/dft.py]: also `CC.pi()` avoids going through symbolic constants; not sure what's better.

File: src/sage/structure/nonexact.py [hunk 9]
```diff
@@ -19,7 +19,7 @@
     sage: R.<x> = PowerSeriesRing(QQ, default_prec=10)
     sage: R.default_prec()
     10
-    sage: cos(x)
+    sage: cos(x)                                                                        # optional - sage.symbolic
```
Comment (orlitzky) [inline src/sage/structure/nonexact.py]: Does using `x.cos()` bypass the symbolic cosine?

Comment (mkoeppe) [reply src/sage/structure/nonexact.py]: Yes.

**Comment by mkoeppe** [comment]
OK, how's this version?

File: src/sage/calculus/calculus.py [hunk 10]
```diff
@@ -935,50 +937,50 @@ def symbolic_product(expression, v, a, b, algorithm='maxima', hold=False):
...
-    OUTPUT: The minimal polynomial of self. If the numerical algorithm
-    is used then it is proved symbolically when epsilon=0 (default).
+    OUTPUT: The minimal polynomial of ``self``. If the numerical algorithm
+    is used, then it is proved symbolically when ``epsilon=0`` (default).
 
     If the minimal polynomial could not be found, two distinct kinds of
     errors are raised. If no reasonable candidate was found with the
-    given bit/degree parameters, a ``ValueError`` will be
+    given ``bit``/``degree`` parameters, a :class:`ValueError` will be
     raised. If a reasonable candidate was found but (perhaps due to
```
Comment (kwankyu) [inline src/sage/calculus/calculus.py:970]: ``bits``

File: src/sage/calculus/functional.py [hunk 11]
```diff
@@ -1,3 +1,4 @@
+# sage.doctest: optional - sage.symbolic
 """
```
Comment (kwankyu) [inline src/sage/calculus/functional.py:2]: `needs sage.symbolic` would work?

File: src/sage/calculus/functions.py [hunk 12]
```diff
@@ -1,9 +1,12 @@
+# sage.doctest: optional - sage.symbolic
 r"""
```
Comment (kwankyu) [inline src/sage/calculus/functions.py:2]: `needs sage.symbolic`

File: src/sage/calculus/integration.pyx [hunk 13]
```diff
@@ -1,3 +1,4 @@
+# sage.doctest: optional - sage.symbolic
 """
```
Comment (kwankyu) [inline src/sage/calculus/integration.pyx:2]: The same.

File: src/sage/calculus/test_sympy.py [hunk 14]
```diff
@@ -1,4 +1,4 @@
-# -*- coding: utf-8 -*-
+# sage.doctest: optional - sage.symbolic
 r"""
```
Comment (kwankyu) [inline src/sage/calculus/test_sympy.py:2]: `needs sage.symbolic`

File: src/sage/calculus/transforms/dft.py [hunk 15]
```diff
@@ -71,20 +71,20 @@
 #
 #                  https://www.gnu.org/licenses/
 ##########################################################################
-from sage.rings.number_field.number_field import CyclotomicField
+from sage.functions.all import sin, cos
 from sage.misc.lazy_import import lazy_import
```
Comment (kwankyu) [inline src/sage/calculus/transforms/dft.py:75]: Is this (importing from `all`) all right?

File: src/sage/calculus/wester.py [hunk 16]
```diff
@@ -1,3 +1,4 @@
+# sage.doctest: optional - sage.symbolic
 r"""
```
Comment (kwankyu) [inline src/sage/calculus/wester.py:2]: `needs sage.symbolic`

File: src/sage/structure/nonexact.py (hunk 9 above)
Comment (kwankyu) [reply src/sage/structure/nonexact.py]: Why not `needs sage.symbolic`?

File: src/sage/calculus/functional.py (hunk 11 above)
Comment (mkoeppe) [reply src/sage/calculus/functional.py:2]: Thanks, fixed (in all files)

File: src/sage/calculus/transforms/dft.py (hunk 15 above)
Comment (mkoeppe) [reply src/sage/calculus/transforms/dft.py:75]: It's not enforced for sage.functions yet, but I've changed it

**Review by kwankyu — APPROVED** [review]
Failed checks seem unrelated with this PR. Otherwise LGTM.

I leave setting positive review to @orlitzky.

**Comment by orlitzky** [comment]
Don't wait for me, I got busy and now all of my sage installs are broken.

**Comment by kwankyu** [comment]
OK. Good wishes for your sage installs!

**Comment by mkoeppe** [comment]
Thanks both!

**Pull request merged**